File: ocaml/bintree.ml
(* bintree.ml
   Binary tree implementation.
   Mostly as a way to test the efficiency of a certain shibby sorting 
   algorithm.
   Turns a list into a binary tree to sort it, then turns the tree back into
   a list.  Wyld, ne?

   Simon Heath
   16/4/2003
 *)


type 'a bintree =
   Empty
 | Node of 'a bintree * 'a * 'a bintree;;

let rec add_node (elt:'a) (tree:'a bintree) =
   match tree with
   Empty -> Node( Empty, elt, Empty )
 | Node( ltree, item, rtree ) ->
      if elt < item then
         Node( add_node elt ltree, item, rtree )
      else
         Node( ltree, item, add_node elt rtree );;

let rec tree_of_list =
   function
   [] -> Empty
 | car :: cdr -> add_node car (tree_of_list cdr);;

let rec list_of_tree =
   function
   Empty -> []
 | Node( ltree, item, rtree ) ->
      (list_of_tree ltree) @ (item :: list_of_tree rtree);;

let sort x = list_of_tree (tree_of_list x);;

File: haskell/bintree.hs
{- bintree.hs
   A binary tree in Haskell.
   Yay!  ^_^

   Simon Heath
   2/7/2003
-}

-- Foist, we declare the data type

module Bintree
   where


data Bintree = 
   Nil 
 | Leaf Int 
 | Node (Bintree) Int (Bintree)


getLeft (Node a b c) = a
getRight (Node a b c) = c
getValue (Node a b c) = b
getValue (Leaf a) = a
getValue Nil = 0

insert :: Int -> Bintree -> Bintree
insert item tree =
   case tree of
      Nil -> Leaf item
      Leaf a -> 
         if item > a
	    then (Node Nil a (Leaf item))
	    else (Node (Leaf item) a Nil)
      Node a b c ->
         if item > b
	    then (insert item c)
	    else (insert item a)
